zrml/swaps/src/fixed.rs

```diff
@@ -180,14 +232,158 @@ mod tests {
+            (10000000000, 1000000000, 10000000000),
+            (10000000000, 2000000000, 10000000000),
+            (10000000000, 3000000000, 10000000000),
+            (10000000000, 4000000000, 10000000000),
+            (10000000000, 5000000000, 10000000000),
+            (10000000000, 6000000000, 10000000000),
+            (10000000000, 7000000000, 10000000000),
+            (10000000000, 8000000000, 10000000000),
+            (10000000000, 9000000000, 10000000000),
+            (10000000000, 10000000000, 10000000000),
+            (12500000000, 0, 10000000000),
+            (12500000000, 1000000000, 10225651825),
+            (12500000000, 2000000000, 10456395525),
+            (12500000000, 3000000000, 10692345999),
+            (12500000000, 4000000000, 10933620739),
+            (12500000000, 5000000000, 11180339887),
+            (12500000000, 6000000000, 11432626298),
+            (12500000000, 7000000000, 11690605597),
+            (12500000000, 8000000000, 11954406247),
+            (12500000000, 9000000000, 12224159606),
+            (12500000000, 10000000000, 12500000000),
+            (15000000000, 0, 10000000000),
+            (15000000000, 1000000000, 10413797439),
+            (15000000000, 2000000000, 10844717711),
+            (15000000000, 3000000000, 11293469354),
+            (15000000000, 4000000000, 11760790225),
+            (15000000000, 5000000000, 12247448713),
+            (15000000000, 6000000000, 12754245006),
+            (15000000000, 7000000000, 13282012399),
+            (15000000000, 8000000000, 13831618672),
+            (15000000000, 9000000000, 14403967511),
+            (15000000000, 10000000000, 15000000000),
+            (17500000000, 0, 10000000000),
+            (17500000000, 1000000000, 10575570503),
+            (17500000000, 2000000000, 11184269147),
+            (17500000000, 3000000000, 11828002689),
+            (17500000000, 4000000000, 12508787635),
+            (17500000000, 5000000000, 13228756555),
+            (17500000000, 6000000000, 13990164762),
+            (17500000000, 7000000000, 14795397379),
+            (17500000000, 8000000000, 15646976811),
+            (17500000000, 9000000000, 16547570643),
+            (17500000000, 10000000000, 17500000000),
+        ];
+        for (base, exp, expected) in test_vector.iter() {
+            let result = bpow_approx(*base, *exp).unwrap();
+            let diff = if result > *expected { result - *expected } else { *expected - result };
+            assert_le!(diff, precision);
+        }
+    }
+
+    #[test]
+    fn bpow_approx_returns_error_when_parameters_are_outside_of_specified_limits() {
+        let test_vector: Vec<(u128, u128)> = vec![
+            (BASE / 10, 3 * BASE / 2),
+            (2 * BASE - BASE / 10, 3 * BASE / 2),
+            (BASE, 11 * BASE / 10),
+        ];
+        for (base, exp) in test_vector.iter() {
+            assert!(bpow_approx(*base, *exp).is_err());
+        }
     }
```

Comment:
Just for the sake of completeness, a test that verifies that the max iteration count check works would be great.

Reply:
Would have loved to do that. Unfortunately, I don't think that's possible. I've profiled the function like this:

```rust
for base in (BASE / 4..7 * BASE / 4).step_by((BASE / 1_000) as usize) {
    for exp in (0..BASE).step_by((BASE / 1_000) as usize) {
        bpow_approx(base, exp).unwrap();
    }
}
```

Maximum number of iterations was 50, and I've got an estimate here that says it should never take more than 70 iterations. So the loop is purely a defense against human error. To me, this is similar to the idea of not having panickers in the code, even if you can argue that they will never trigger.

If you think it's worth it, I can do the following: Put the `precision` parameter back in and then test that with sufficiently small `precision` the errors gets triggered.

Reply:
Thanks for the explanation. We can leave it like that.